protocol/messenger_communities.go

```diff
@@ -1921,3 +1914,389 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+
+		// We're calling `CreateCommunity` on `communitiesManager` directly, instead of
+		// using the `Messenger` API, so we get more control over when we set up filters,
+		// the community is published and data is being synced (we don't want the community
+		// to show up in clients while the import is in progress)
+		discordCommunity, err := m.communitiesManager.CreateCommunity(createCommunityRequest, false)
+		if err != nil {
+			importProgress.AddTaskError(discord.CommunityCreationTask, discord.Error(err.Error()))
+			importProgress.StopTask(discord.CommunityCreationTask)
+			m.publishImportProgress(importProgress)
+			return
+		}
+
+		communitySettings := communities.CommunitySettings{
+			CommunityID:                  discordCommunity.IDString(),
+			HistoryArchiveSupportEnabled: true,
+		}
+		err = m.communitiesManager.SaveCommunitySettings(communitySettings)
+		if err != nil {
+			m.cleanUpFailedImport(discordCommunity)
+			importProgress.AddTaskError(discord.CommunityCreationTask, discord.Error(err.Error()))
+			importProgress.StopTask(discord.CommunityCreationTask)
+			m.publishImportProgress(importProgress)
+			return
+		}
+
+		importProgress.CommunityID = discordCommunity.IDString()
+		importProgress.UpdateTaskProgress(discord.CommunityCreationTask, 0.75)
+		m.publishImportProgress(importProgress)
+
+		//This is a map of discord category IDs <-> Status category IDs
+		processedCategoriesIds := make(map[string]string, 0)
+		totalCategoriesCount := len(exportData.Categories)
+
+		for _, category := range exportData.Categories {
+
+			createCommunityCategoryRequest := &requests.CreateCommunityCategory{
+				CommunityID:  discordCommunity.ID(),
+				CategoryName: category.Name,
+				ChatIDs:      make([]string, 0),
+			}
+			processedCategoriesIds[category.ID] = ""
+			progressValue := (float32(len(processedCategoriesIds)) / float32(totalCategoriesCount))
+			importProgress.UpdateTaskProgress(discord.CommunityCreationTask, 0.75+(progressValue*0.25))
+
+			// We call `CreateCategory` on `communitiesManager` directly so we can control
+			// whether or not the community update should be published (it should not until the
+			// import has finished)
+			communityWithCategories, changes, err := m.communitiesManager.CreateCategory(createCommunityCategoryRequest, false)
+			if err != nil {
+				m.cleanUpFailedImport(discordCommunity)
+				importProgress.AddTaskError(discord.CommunityCreationTask, discord.Error(err.Error()))
+				importProgress.StopTask(discord.CommunityCreationTask)
+				m.publishImportProgress(importProgress)
+				return
+			}
+			discordCommunity = communityWithCategories
+
+			// This looks like we keep overriding the same field but there's
+			// only one `CategoriesAdded` change at this point.
+			for _, addedCategory := range changes.CategoriesAdded {
+				processedCategoriesIds[category.ID] = addedCategory.CategoryId
+			}
+			m.publishImportProgress(importProgress)
+		}
```

Comment:
Notice how the categories are created sequentially. Theoretically, we can do this in parallel, but judging by what I've experienced using real data, this is rather fast. Anyways, if anyone thinks it's worth it to pursue I'm happy to do that in a follow up PR. Right now I don't see the need.